### DISCUSSION TITLE
Java 8 Compatibility

### DISCUSSION THREAD

**Discussion by liammacisaac** [ebourg/jsign#223] [Q&A] [answered]
Are you looking to keep JSign compatible with Java 8 in the 6.1 release? I've noticed that there's something breaking in there which works fine with Java 21, specifically when using microsoft trusted signing. The pom still has source/target = 8 though which makes me think that this is an oversight.

If you can confirm that you want to keep Java 8 support in this version, I'll raise an issue, investigate a bit further and submit a PR to fix it.

**Comment by ebourg** [comment]
Yes I intend to preserve the Java 8 compatibility, ideally until Java 8 is EOL everywhere. What issue did you notice?

**Reply by liammacisaac** [reply]
I didn't have much time to look into the issue, but taking a look now, I've realised it's user error. My IntelliJ had defaulted to JDK21 for the project so I was running code compiled on JDK 21 with a V8 JRE. The main class that causes an issue here is `ByteBuffer` because some of the methods that previously inherited from `Buffer` (e.g. `flip()` in this case) now have overridden implementations in `ByteBuffer` that return a `ByteBuffer` object. This then causes JDK8 to throw a NoSuchMethodError at runtime. So, as long as it's compiled with Java 8 there's no issue with compatibility from 8-21.

**Reply by ebourg** [reply, marked as answer]
Ah yes that's a common issue, it could be avoided by building with the `-release 8` flag but that doesn't work for Jsign. With this option the compiler no longer recognizes the javax.smartcardio package used for the PIV and OpenPGP cards ([JDK-8212638](https://bugs.openjdk.org/browse/JDK-8212638)), nor the internal packages such as sun.security.pkcs11.wrapper which is used for configuring the PKCS#11 libraries.

I don't mind adding some casts around the ByteBuffer calls to avoid this issue.

**Reply by liammacisaac** [reply]
Just to wrap this up, I'm OK with how this works at the moment now knowing what I did wrong.
